runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java

```diff
@@ -143,32 +170,44 @@ private PTransformTranslation() {}
         DisplayDataTranslation.toProto(DisplayData.from(appliedPTransform.getTransform())));
 
     PTransform<?, ?> transform = appliedPTransform.getTransform();
+
     // A RawPTransform directly vends its payload. Because it will generally be
     // a subclass, we cannot do dictionary lookup in KNOWN_PAYLOAD_TRANSLATORS.
     if (transform instanceof RawPTransform) {
-      RawPTransform<?, ?> rawPTransform = (RawPTransform<?, ?>) transform;
-
-      if (rawPTransform.getUrn() != null) {
-        FunctionSpec.Builder payload = FunctionSpec.newBuilder().setUrn(rawPTransform.getUrn());
-        @Nullable ByteString parameter = rawPTransform.getPayload();
-        if (parameter != null) {
-          payload.setPayload(parameter);
-        }
-        transformBuilder.setSpec(payload);
-      }
-      rawPTransform.registerComponents(components);
+      // The raw transform was parsed in the context of other components; this puts it in the
+      // context of our current serialization
+      transformBuilder.setSpec(((RawPTransform<?, ?>) transform).migrate(components));
     } else if (KNOWN_PAYLOAD_TRANSLATORS.containsKey(transform.getClass())) {
-      FunctionSpec payload =
+      transformBuilder.setSpec(
           KNOWN_PAYLOAD_TRANSLATORS
               .get(transform.getClass())
-              .translate(appliedPTransform, components);
-      transformBuilder.setSpec(payload);
+              .translate(appliedPTransform, components));
     }
 
     return transformBuilder.build();
   }
 
   /**
+   * Translates a {@link RunnerApi.PTransform} to a {@link RawPTransform} specialized for the URN
+   * and spec.
+   */
+  static RawPTransform<?, ?> rehydrate(
+      RunnerApi.PTransform protoTransform, RehydratedComponents rehydratedComponents)
+      throws IOException {
+
+    @Nullable
+    TransformPayloadTranslator<?> rehydrator =
+        KNOWN_REHYDRATORS.get(
+            protoTransform.getSpec() == null ? null : protoTransform.getSpec().getUrn());
+
+    if (rehydrator == null) {
+      return UnknownRawPTransform.withSpec(protoTransform.getSpec());
```

Comment:
<!--new_thread; commit:042367c660b607e55c9faafa4d3d8a8685547151; resolved:0-->
I kind of like just having a default rehydrator that we use (given that there's a method on the Translator already). Up to you.